### ISSUE TITLE
mini-batch GraphSAGE training: drop in test accuracy for power of 2 batch sizes

### ISSUE THREAD

**Issue by yuk12** [dmlc/dgl#3906]
## 🐛 Bug

We observe that the training accuracy drops significantly when we choose mini-batch size around power of 2.
For example: there is a significant drop in accuracy between 4000 and 4096 batch-size, or (33000 or 32000) and 32768 batch-size.
Such difference in accuracy is stark at 32K batch-size till 4k batch size. The difference is not distinctly clear for 2K and 1K batch sizes.
We did experiments with a range of batch sizes and here are the results.
batch size range: 33K to 32K:

<html xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=PowerPoint.Slide>
<meta name=Generator content="Microsoft PowerPoint 15">
<style>

</style>
</head>

<body>

Epoch/batch_size | 33000 | 32768 | 32766 | 32764 | 32762 | 32760 | 32750 | 32740 | 32730 | 32720 | 32710 | 32700 | 32650 | 32600 | 32550 | 32500 | 32450 | 32400 | 32350 | 32300 | 32250 | 32200 | 32150 | 32100 | 32050 | 32000
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
5 | 0.6777 | 0.6426 | 0.6706 | 0.6626 | 0.6722 | 0.6804 | 0.6758 | 0.6833 | 0.6856 | 0.6802 | 0.6798 | 0.6747 | 0.6875 | 0.6897 | 0.6865 | 0.6889 | 0.6905 | 0.6862 | 0.6947 | 0.6881 | 0.694 | 0.6902 | 0.6858 | 0.6964 | 0.6913 | 0.6941
10 | 0.7088 | 0.6715 | 0.6882 | 0.6908 | 0.6944 | 0.7086 | 0.7036 | 0.7111 | 0.7094 | 0.7133 | 0.7075 | 0.7065 | 0.7144 | 0.7228 | 0.7182 | 0.7155 | 0.7161 | 0.7152 | 0.7235 | 0.7192 | 0.7215 | 0.7188 | 0.7149 | 0.7217 | 0.7239 | 0.7198
15 | 0.7276 | 0.7016 | 0.7057 | 0.7008 | 0.6978 | 0.7098 | 0.7169 | 0.7236 | 0.722 | 0.7271 | 0.7242 | 0.7239 | 0.7299 | 0.7308 | 0.7301 | 0.7315 | 0.7307 | 0.731 | 0.7382 | 0.7339 | 0.7378 | 0.7367 | 0.7286 | 0.7381 | 0.7399 | 0.7365
20 | 0.7381 | 0.707 | 0.7077 | 0.7071 | 0.707 | 0.7177 | 0.7187 | 0.7275 | 0.7335 | 0.7361 | 0.736 | 0.732 | 0.7385 | 0.7455 | 0.7388 | 0.7439 | 0.7446 | 0.7414 | 0.7509 | 0.7489 | 0.752 | 0.7463 | 0.7371 | 0.7523 | 0.7487 | 0.7499
25 | 0.7483 | 0.707 | 0.7101 | 0.7165 | 0.7198 | 0.7223 | 0.7277 | 0.7395 | 0.7411 | 0.7411 | 0.7426 | 0.7393 | 0.7559 | 0.7566 | 0.7459 | 0.7559 | 0.7555 | 0.7559 | 0.7652 | 0.7584 | 0.7635 | 0.7634 | 0.7499 | 0.7616 | 0.7633 | 0.7658
30 | 0.7567 | 0.707 | 0.7114 | 0.7198 | 0.7257 | 0.716 | 0.7326 | 0.7421 | 0.7497 | 0.7506 | 0.747 | 0.7508 | 0.7572 | 0.7633 | 0.7561 | 0.7643 | 0.7573 | 0.7604 | 0.7711 | 0.7694 | 0.7718 | 0.7671 | 0.7592 | 0.7712 | 0.7685 | 0.7701
35 | 0.766 | 0.7098 | 0.7159 | 0.7206 | 0.7401 | 0.7223 | 0.7326 | 0.7478 | 0.747 | 0.7487 | 0.7606 | 0.7592 | 0.7682 | 0.768 | 0.7679 | 0.7708 | 0.7677 | 0.773 | 0.7779 | 0.772 | 0.773 | 0.7702 | 0.7709 | 0.7779 | 0.7734 | 0.7786
40 | 0.7704 | 0.7098 | 0.7159 | 0.7265 | 0.7345 | 0.736 | 0.7325 | 0.7478 | 0.747 | 0.7602 | 0.7569 | 0.7621 | 0.7738 | 0.7708 | 0.7679 | 0.7732 | 0.7704 | 0.7765 | 0.7784 | 0.7781 | 0.776 | 0.7769 | 0.7762 | 0.7788 | 0.7805 | 0.7769

</body>

</html>

batch size: 8K
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/mvasimud/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/mvasimud/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

Epoch/batch_size | 8192 | 8000
-- | -- | --
5 | 0.7153 | 0.7425
10 | 0.7388 | 0.7661
15 | 0.745 | 0.7805
20 | 0.7694 | 0.7849
25 | 0.7532 | 0.7882
30 | 0.7652 | 0.7866
35 | 0.7698 | 0.7905
40 | 0.7704 | 0.7938

</body>

</html>

batch size range: 4K - 5K:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/mvasimud/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/mvasimud/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

Epoch/batch_size | 4000 | 4010 | 4020 | 4030 | 4040 | 4050 | 4060 | 4070 | 4080 | 4090 | 4096 | 4100 | 4200 | 4300 | 4400 | 4500 | 4600 | 4700 | 4800 | 4900 | 5000 | 4096
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
5 | 0.7561 | 0.7582 | 0.7594 | 0.7475 | 0.7502 | 0.747 | 0.7632 | 0.7543 | 0.759 | 0.7558 | 0.7466 | 0.76 | 0.7515 | 0.7532 | 0.7669 | 0.7669 | 0.7551 | 0.7441 | 0.7394 | 0.7538 | 0.7498 | 0.7466
10 | 0.7685 | 0.7673 | 0.7757 | 0.7796 | 0.7746 | 0.7759 | 0.7801 | 0.775 | 0.7841 | 0.7793 | 0.7698 | 0.7761 | 0.7763 | 0.7778 | 0.7847 | 0.7847 | 0.7753 | 0.7746 | 0.762 | 0.7759 | 0.7718 | 0.7698
15 | 0.7808 | 0.7848 | 0.7878 | 0.7896 | 0.7823 | 0.7832 | 0.7876 | 0.7835 | 0.7781 | 0.7793 | 0.7662 | 0.7814 | 0.7797 | 0.7855 | 0.7871 | 0.7871 | 0.7798 | 0.7872 | 0.7786 | 0.7842 | 0.7831 | 0.7662
20 | 0.7892 | 0.7858 | 0.7876 | 0.7854 | 0.7837 | 0.7876 | 0.7909 | 0.7883 | 0.7833 | 0.7845 | 0.77 | 0.7923 | 0.783 | 0.7887 | 0.7859 | 0.7859 | 0.7889 | 0.7888 | 0.7812 | 0.7891 | 0.7866 | 0.77
25 | 0.788 | 0.7907 | 0.7914 | 0.7923 | 0.7879 | 0.7915 | 0.7932 | 0.7953 | 0.7928 | 0.7891 | 0.7792 | 0.7938 | 0.7898 | 0.7933 | 0.7935 | 0.7935 | 0.7871 | 0.7914 | 0.7898 | 0.7901 | 0.7894 | 0.7792
30 | 0.7917 | 0.7898 | 0.7942 | 0.7904 | 0.7939 | 0.7973 | 0.7908 | 0.7934 | 0.7926 | 0.7863 | 0.7739 | 0.7941 | 0.7922 | 0.7922 | 0.7944 | 0.7944 | 0.7888 | 0.7885 | 0.7947 | 0.7912 | 0.7916 | 0.7739
35 | 0.7896 | 0.7924 | 0.7963 | 0.7937 | 0.7909 | 0.795 | 0.7942 | 0.7994 | 0.7909 | 0.7909 | 0.7638 | 0.7913 | 0.7892 | 0.7902 | 0.7965 | 0.7965 | 0.7922 | 0.7953 | 0.7938 | 0.7949 | 0.7925 | 0.7638
40 | 0.7887 | 0.7906 | 0.7952 | 0.7893 | 0.7926 | 0.7951 | 0.7949 | 0.7962 | 0.7917 | 0.7916 | 0.7844 | 0.7841 | 0.791 | 0.7935 | 0.7881 | 0.7881 | 0.7945 | 0.7943 | 0.7924 | 0.7907 | 0.7969 | 0.7844

</body>

</html>

 
## To Reproduce
GNN model: GraphSAGE mini-batch training (3 layers, 256 hidden size, fan-out: 5,10,15 , dropout: 0.50)
Dataset: ogbn-products
lr = 0.06 for 33K - 32K batch-size
lr = 0.03 for 4K- 5K batch-size

1.
1.
1.

## Environment

 - DGL Version (e.g., 1.0): v0.7.2
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): PyTorch v1.7.1
 - OS (e.g., Linux): Linux
 - How you installed DGL (`conda`, `pip`, source): Source 
 - Build command you used (if compiling from source): 'make -j'    (gcc 8.3.0)
 - Python version: 3.8
 - CUDA/cuDNN version (if applicable):
 - GPU models and configuration (e.g. V100):
 - Any other relevant information:

**Comment by BarclayII** [comment]
Does it still happen in 0.8.0post2?

**Comment by yuk12** [comment]
We did not get a chance to try it on 0.8 release. We use commit: 582f71a (v0.7.2) to do these experiments.
